plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/RewriteCast.java

```diff
@@ -0,0 +1,99 @@
+import io.trino.plugin.jdbc.expression.AbstractRewriteCast;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.type.CharType;
+import io.trino.spi.type.Type;
+import io.trino.spi.type.VarcharType;
+import oracle.jdbc.OracleTypes;
+
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+import static io.trino.plugin.oracle.OracleClient.ORACLE_CHAR_MAX_CHARS;
+import static io.trino.plugin.oracle.OracleClient.ORACLE_VARCHAR2_MAX_CHARS;
+
+public class RewriteCast
+        extends AbstractRewriteCast
+{
+    public RewriteCast(BiFunction<ConnectorSession, Type, String> toTargetType)
+    {
+        super(toTargetType);
+    }
+
+    @Override
+    protected boolean eligibleForPushdown(JdbcTypeHandle sourceType, Type targetType)
+    {
+        if (targetType instanceof CharType charType) {
+            // Oracle will throw an error on casts to char(n>ORACLE_CHAR_MAX_CHARS) "ORA-00932: inconsistent datatypes: expected - got NCLOB"
+            return charType.getLength() <= ORACLE_CHAR_MAX_CHARS
+                    && supportedSourceTypeToCastInChar(sourceType);
+        }
+        if (targetType instanceof VarcharType varcharType) {
+            return !varcharType.isUnbounded()
+                    // unbounded varchar and char(n>ORACLE_VARCHAR2_MAX_CHARS) gets written as nclob.
+                    // pushdown does not happen when comparing nclob type variable, so skipping to pushdown the cast for nclob type variable.
+                    && varcharType.getLength().orElseThrow() <= ORACLE_VARCHAR2_MAX_CHARS
+                    && supportedSourceTypeToCastInVarchar(sourceType);
+        }
+        return false;
+    }
+
+    @Override
+    protected Optional<JdbcTypeHandle> toJdbcTypeHandle(Type type)
+    {
+        if (type instanceof CharType charType) {
+            return Optional.of(new JdbcTypeHandle(OracleTypes.CHAR, Optional.of(charType.getBaseName()), Optional.of(charType.getLength()), Optional.empty(), Optional.empty(), Optional.empty()));
+        }
+        if (type instanceof VarcharType varcharType) {
+            return Optional.of(new JdbcTypeHandle(OracleTypes.VARCHAR, Optional.of(varcharType.getBaseName()), varcharType.getLength(), Optional.empty(), Optional.empty(), Optional.empty()));
+        }
+        return Optional.empty();
+    }
+
+    private static boolean supportedSourceTypeToCastInVarchar(JdbcTypeHandle sourceType)
+    {
+        return switch (sourceType.jdbcType()) {
+            case OracleTypes.NUMBER,
+                    OracleTypes.TINYINT,
+                    OracleTypes.SMALLINT,
+                    OracleTypes.INTEGER,
+                    OracleTypes.BIGINT,
+                    OracleTypes.DECIMAL,
+                    OracleTypes.CHAR,
+                    OracleTypes.VARCHAR,
+                    OracleTypes.NCHAR,
+                    OracleTypes.NVARCHAR,
+                    OracleTypes.CLOB,
+                    OracleTypes.NCLOB -> true;
```

Comment:
Since NCLOB doesn't support pushdown - does casting into varchar is being supported by Oracle - Does it break the existing semantics we have today ?

Reply:
`nclob` as the source type can be type cast in `varchar` type. But otherway around is not possible directly using `CAST`.

```
SQL> select cast(c_nclob AS varchar2(10 char)) from presto_test_user.table_1_dn5kb0yskx;

CAST(C_NCL
----------
India
Poland


SQL> select cast(C_VARCHAR_10 AS nclob) from presto_test_user.table_1_dn5kb0yskx;
select cast(C_VARCHAR_10 AS nclob) from presto_test_user.table_1_dn5kb0yskx
       *
ERROR at line 1:
ORA-00932: inconsistent datatypes: expected - got NCLOB
```

Reply:
NCLOB does case insensitive search right ? So in that case can we have a test case wrt to join which ensures join condition is not applied in a case insensitive way